File: Example/Closure/closure_practise.py
from typing import Tuple, Callable


def create_counter(init: int) -> Tuple[Callable[..., int], Callable[..., int], Callable[..., int]]:
    cur_value = init

    def add():
        nonlocal cur_value
        cur_value += 1
        return cur_value

    def dec():
        nonlocal cur_value
        cur_value -= 1
        return cur_value

    def reset():
        nonlocal cur_value
        cur_value = init
        return cur_value
    return add, dec, reset

if __name__ == '__main__':
    add, dec, reset = create_counter(5)
    print(add())
    print(reset())
    print(dec())

# 6
# 5
# 4

File: Example/Closure/ClosurePractise.java
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

public class ClosurePractise {

    private int curValue;
    public Supplier[] createCounter(int init) {
        this.curValue = init;
        AtomicInteger local = new AtomicInteger(init + 1);
        Supplier<int[]> add = () -> new int[]{++this.curValue, local.incrementAndGet()};
        Supplier<int[]> dec = () -> new int[]{--this.curValue, local.decrementAndGet()};
        Supplier<int[]> reset = () -> {
            this.curValue = init;
            local.set(init + 1);
            return new int[]{this.curValue, local.get()};
        };
        return new Supplier[] {add, dec, reset};
    }

    public static void main(String[] args) {
        var sol = new ClosurePractise();
        var counters = sol.createCounter(5);
        var res = new int[][] {(int [])counters[0].get(), (int [])counters[2].get(), (int [])counters[1].get()};
        for (var unit: res) {
            System.out.printf("%d %d\n", unit[0], unit[1]);
        }
    }
}
// 6 7
// 5 6
// 4 5
